/*
 * CuadraAiLib
 *
 * This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
 */

package ai.cuadra.api.controllers;

import ai.cuadra.api.ApiHelper;
import ai.cuadra.api.Server;
import ai.cuadra.api.exceptions.ApiException;
import ai.cuadra.api.exceptions.ErrorResponseException;
import ai.cuadra.api.http.request.HttpMethod;
import ai.cuadra.api.http.response.ApiResponse;
import ai.cuadra.api.models.Chat;
import ai.cuadra.api.models.ChatResponseEx;
import io.apimatic.core.ApiCall;
import io.apimatic.core.ErrorCase;
import io.apimatic.core.GlobalConfiguration;
import io.apimatic.coreinterfaces.http.request.ResponseClassType;
import java.io.IOException;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;

/**
 * This class lists all the endpoints of the groups.
 */
public final class ChatController extends BaseController {

    /**
     * Initializes the controller.
     * @param globalConfig    Configurations added in client.
     */
    public ChatController(GlobalConfiguration globalConfig) {
        super(globalConfig);
    }

    /**
     * This endpoint creates a new chat interaction with our AI Models.
     * @param  body  Required parameter:
     * @return    Returns the ChatResponseEx wrapped in ApiResponse response from the API call
     * @throws    ApiException    Represents error response from the server.
     * @throws    IOException    Signals that an I/O exception of some sort has occurred.
     */
    public ApiResponse<ChatResponseEx> chat(
            final Chat body) throws ApiException, IOException {
        return prepareChatRequest(body).execute();
    }

    /**
     * This endpoint creates a new chat interaction with our AI Models.
     * @param  body  Required parameter:
     * @return    Returns the ChatResponseEx wrapped in ApiResponse response from the API call
     */
    public CompletableFuture<ApiResponse<ChatResponseEx>> chatAsync(
            final Chat body) {
        try {
            return prepareChatRequest(body).executeAsync();
        } catch (Exception e) {
            throw new CompletionException(e);
        }
    }

    /**
     * Builds the ApiCall object for chat.
     */
    private ApiCall<ApiResponse<ChatResponseEx>, ApiException> prepareChatRequest(
            final Chat body) {
        return new ApiCall.Builder<ApiResponse<ChatResponseEx>, ApiException>()
                .globalConfig(getGlobalConfiguration())
                .requestBuilder(requestBuilder -> requestBuilder
                        .server(Server.ENUM_DEFAULT.value())
                        .path("/chat")
                        .bodyParam(param -> param.value(body))
                        .bodySerializer(() ->  ApiHelper.serialize(body))
                        .headerParam(param -> param.key("Content-Type")
                                .value("application/json").isRequired(false))
                        .headerParam(param -> param.key("accept").value("application/json"))
                        .withAuth(auth -> auth
                                .add("OAuth2"))
                        .httpMethod(HttpMethod.POST))
                .responseHandler(responseHandler -> responseHandler
                        .responseClassType(ResponseClassType.API_RESPONSE)
                        .apiResponseDeserializer(
                                response -> ApiHelper.deserialize(response, ChatResponseEx.class))
                        .nullify404(false)
                        .localErrorCase("400",
                                 ErrorCase.setReason("Bad request, read again our documentation or contact support for guidance.",
                                (reason, context) -> new ErrorResponseException(reason, context)))
                        .localErrorCase("401",
                                 ErrorCase.setReason("Not authorized, check our OAuth2 doc.",
                                (reason, context) -> new ErrorResponseException(reason, context)))
                        .localErrorCase("500",
                                 ErrorCase.setReason("Internal error, if this error persist, please contact support.",
                                (reason, context) -> new ErrorResponseException(reason, context)))
                        .globalErrorCase(GLOBAL_ERROR_CASES))
                .build();
    }
}